django/core/files/storage.py

```diff
@@ -415,6 +423,220 @@ def get_modified_time(self, name):
+
+
+class InMemoryDirNode:
+    """
+    Helper class representing an in-memory directory node
+    """
+
+    def __init__(self):
+        self._children = {}
+
+    def resolve(self, path, create_if_missing=False, leaf_cls=None, check_exists=True):
+        """
+        Navigate current directory tree, returning node matching path or
+        creating it if missing.
+
+        Args:
+            path: path of the node to search
+            create_if_missing: create nodes if they don't exist. Defaults to False.
+            leaf_cls: expected type of leaf node. Defaults to None.
+            check_exists: if true and the leaf node does not exist, raise
+                          a FileNotFoundError. Defaults to True.
+        """
+        path_segments = path.split("/")
+        current_node = self
+
+        while path_segments:
+            path_segment = path_segments.pop(0)
+
+            # If current node is a file node and there are unprocessed segments,
+            # raise an error
+            if isinstance(current_node, InMemoryFileNode):
+                path_segments = path.split("/")
+                current_path = "/".join(
+                    path_segments[: path_segments.index(path_segment)]
+                )
+                raise NotADirectoryError(
+                    errno.ENOTDIR, os.strerror(errno.ENOTDIR), current_path
+                )
+
+            current_node = current_node._resolve_child(
+                path_segment,
+                create_if_missing,
+                leaf_cls if len(path_segments) == 0 else InMemoryDirNode,
+            )
+            if current_node is None:
+                break
+
+        if current_node is None and check_exists:
+            raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), path)
+
+        # If a leaf_cls is not None, check if leaf node is of right type
+        if leaf_cls and not isinstance(current_node, leaf_cls):
+            error_cls, error_code = (
+                (NotADirectoryError, errno.ENOTDIR)
+                if leaf_cls is InMemoryDirNode
+                else (IsADirectoryError, errno.EISDIR)
+            )
+            raise error_cls(error_code, os.strerror(error_code), path)
+
+        return current_node
+
+    def _resolve_child(self, path_segment, create_if_missing, child_cls):
+        if path_segment in ("", "."):
+            return self
+        if create_if_missing:
+            return self._children.setdefault(path_segment, child_cls())
+        return self._children.get(path_segment)
+
+    def list_dir_names(self):
+        return [k for k, v in self._children.items() if isinstance(v, InMemoryDirNode)]
+
+    def list_file_names(self):
+        return [k for k, v in self._children.items() if isinstance(v, InMemoryFileNode)]
+
+    def remove_child(self, name):
+        if name in self._children:
+            del self._children[name]
+
+
+@deconstructible()
+class InMemoryStorage(Storage, StorageSettingsMixin):
```

Comment:
Could it be useful to have each Storage implementation in a separate file? So how does it happen with the different Cache backends?

Reply:
Indeed... It was actually my first implementation, then I reverted to a single-file module to keep tests about deconstruction working. I didn't check `@deconstructible` docs, my fault 😅